crates/sui-metrics-proxy/src/admin.rs

```diff
@@ -0,0 +1,158 @@
+    info!(
+        "signal received, starting graceful shutdown, grace period {} seconds, if needed",
+        &grace
+    );
+    h.graceful_shutdown(Some(Duration::from_secs(grace)))
+}
+
+/// App will configure our routes and create our mpsc channels.  This fn is also used to instrument
+/// our tests
+pub fn app(buffer_size: usize, network: String) -> Router {
+    // we accept data on our UpstreamConsumer up to our buffer size.
+    let (sender, receiver) = mpsc::channel(buffer_size);
+    let mut consumer = UpstreamConsumer::new(network, receiver);
+
+    tokio::spawn(async move { consumer.run().await });
+
+    // build our application with a route and our sender mpsc
+    Router::new()
+        .route("/publish/metrics", axum_post(publish_metrics))
+        .with_state(Arc::new(sender))
+}
+
+pub async fn server(
+    listener: std::net::TcpListener,
+    _acceptor: TlsAcceptor, // TODO enable for tls
+    app: Router,
+) -> std::io::Result<()> {
+    // setup our graceful shutdown
+    let handle = axum_server::Handle::new();
+    // Spawn a task to gracefully shutdown server.
+    tokio::spawn(shutdown_signal(handle.clone()));
+
+    axum_server::Server::from_tcp(listener)
+        // .acceptor(acceptor) // TODO enable for tls
+        .handle(handle)
+        .serve(app.into_make_service())
+        .await
+}
+
+pub fn create_server_cert(
+    hostname: &str,
+) -> Result<(ServerConfig, ValidatorAllowlist), sui_tls::rustls::Error> {
+    let mut rng = rand::thread_rng();
+    let server_keypair = Ed25519KeyPair::generate(&mut rng);
+    let server_certificate = SelfSignedCertificate::new(server_keypair.private(), hostname);
+
+    ValidatorCertVerifier::rustls_server_config(
+        vec![server_certificate.rustls_certificate()],
+        server_certificate.rustls_private_key(),
+    )
+}
+
+async fn get_validators(url: String) -> Result<Vec<ValidatorMetadata>> {
+    let client = reqwest::Client::builder().build().unwrap();
+    let request = serde_json::json!({
+        "jsonrpc": "2.0",
+        "method":"sui_getValidators",
+        "id":1,
+    });
+    let response = client
+        .post(url)
+        .header(reqwest::header::CONTENT_TYPE, "application/json")
+        .body(request.to_string())
+        .send()
+        .await
+        .context("unable to perform rpc")?;
+
+    #[derive(Debug, Deserialize)]
+    struct ResponseBody {
+        result: Vec<ValidatorMetadata>,
+    }
+    let body = response
+        .json::<ResponseBody>()
+        .await
+        .context("unable to deserialize validator peer list")?;
+
+    Ok(body.result)
+}
+
+pub fn manage_validators(url: String, period: Duration, allowlist: ValidatorAllowlist) {
+    info!("Started polling for peers using rpc: {}", url);
+    tokio::spawn(async move {
+        let mut interval = tokio::time::interval(period);
+        interval.set_missed_tick_behavior(tokio::time::MissedTickBehavior::Skip);
+
+        loop {
+            interval.tick().await;
```

Comment:
note that if the interval is too wide, we could miss the epoch boundary and cause loss of some metrics due to validator joining and metadata rotation. Same thing for the depended fullnode. Do we have a plan how to mitigate this potential single point of failure?

Reply:
this interval is configurable from the cli args.  No strategy other than the push metrics are best effort delivery.